### PR TITLE
fix(es-rollover): parameterize archive template mappings and add inte…

### LINKED ISSUE #8292: [Bug]: jaegertracing 2.17 creates new index instead of using write alias with ILM on elastic-search
### What happened?

For the past 2 days I've tried to enable archive storage on elastic-search and had quite troubles.

The policies are created and somewhat assigend, but it keeps creating `jaeger-archive-jaeger-span-write`  index instead of writing to the alias and the actual documents going to the rollover index `jaeger-archive-jaeger-span-archive-000001`

### Steps to reproduce

Prepare ilm policy

```bash
# regular ILM Policy
docker run --rm --net=host \
  -e ES_USERNAME=jaeger \
  -e ES_PASSWORD=<password> \
  -e ES_USE_ILM=true \
  jaegertracing/jaeger-es-rollover:latest \
  init https://host.docker.internal:9200 \
  --es.tls.enabled=true \
  --es.tls.skip-host-verify=true \
  --shards=1 \
  --replicas=1

# for archive
 docker run --rm --net=host \
  -e ES_USERNAME=jaeger \
  -e ES_PASSWORD=<password> \
  -e ES_ILM_POLICY_NAME=jaeger-ilm-policy-archive \
  jaegertracing/jaeger-es-rollover:latest \
  init https://host.docker.internal:9200 \
  --es.tls.enabled=true \
  --es.tls.skip-host-verify=true \
  --archive=true \
  --index-prefix=jaeger-archive \
  --shards=1 \
  --replicas=1 \
  --es.ilm-policy-name=jaeger-ilm-policy-archive \
  --es.use-ilm=true
```

CustomResource:

```yaml

apiVersion: opentelemetry.io/v1beta1
kind: OpenTelemetryCollector
metadata:
  name: jaeger-telemetry
spec:
  image: jaegertracing/jaeger:2.17.0
  ports:
  - name: jaeger
    port: 16686
  env:
  - name: ES_USERNAME
    valueFrom:
      secretKeyRef:
        name: jaeger-es-credentials
        key: username
  - name: ES_PASSWORD
    valueFrom:
      secretKeyRef:
        name: jaeger-es-credentials
        key: password
  resources:
    limits:
      cpu: "1"
      memory: "2Gi"
    requests:
      cpu: "100m"
      memory: "32Mi"
  volumes:
  - name: trust-bundle
    configMap:
      name: trust-bundle
      items:
        - key: trust-bundle.pem
          path: trust-bundle.pem 
  volumeMounts:
  - name: trust-bundle
    mountPath: /etc/ssl/certs/trust-bundle.pem
    subPath: trust-bundle.pem
    readOnly: true
  config:
    service:
      extensions: [jaeger_storage, jaeger_query]
      pipelines:
        traces:
          receivers: [otlp]
          exporters: [jaeger_storage_exporter]
    extensions:
      jaeger_query:
        storage:
          traces: elastic_storage
          traces_archive: elastic_storage_archive
      jaeger_storage:
        backends:
          elastic_storage: &elasticstorage_config
            elasticsearch:
              server_urls:
                - https://elk-default-es-http.monitoring.svc:9200
              auth:
                basic:
                  username: "${ES_USERNAME}"
                  password: "${ES_PASSWORD}"
              max_span_age: 2880h  # 120 days (120 * 24 hours)
              create_mappings: false 
              use_ilm: true
              use_aliases: true 
          elastic_storage_archive:
            elasticsearch:
              server_urls:
                - https://elk-default-es-http.monitoring.svc:9200
              auth:
                basic:
                  username: "${ES_USERNAME}"
                  password: "${ES_PASSWORD}"
              indices:
                index_prefix: "jaeger-archive"
              create_mappings: false
              use_ilm: true
              use_aliases: true
			  # These below also didn't help
              # span_write_alias: jaeger-archive-jaeger-span-write
              # span_read_alias: jaeger-archive-jaeger-span-read
              # service_write_alias: jaeger-archive-jaeger-servce-write
              # service_read_alias: jaeger-archive-jaeger-servce-read
        metric_backends:
          elastic_storage: *elasticstorage_config
    receivers:
      otlp:
        protocols:
          grpc:
            endpoint: 0.0.0.0:4317
          http:
            endpoint: 0.0.0.0:4318
    exporters:
      jaeger_storage_exporter:
        trace_storage: elastic_storage
```

After deployment and making a request, and looking at the trace, archive it. 

Image

it completley ignores the `create_mapping: false` (or use_aliases)  and creates a new index `jaeger-archive-jaeger-span-write` and writes into it, instead of `jaeger-archive-jaeger-span-000001` via rollover alias. 

The ILM looks fine

```json
{
  "index": {
    "lifecycle": {
      "name": "jaeger-ilm-policy-archive",
      "rollover_alias": "jaeger-archive-jaeger-span-write"
    },
    "requests": {
      "cache": {
        "enable": "true"
      }
    },
    "number_of_shards": "1",
    "number_of_replicas": "1",
    "mapping": {
      "nested_fields": {
        "limit": "50"
      }
    }
  }
}
```

For the regular primary storage it works fine and it writes to the `-000001` rollovers.

### Expected behavior

Archived spans written to `jaeger-archive-jaeger-span-archive-000001`. 

### Relevant log output

```shell
No logs, except sometimes after repeating this steps I'd get a index rollover warning in Kibana with these logs, which I can "retry" from Kibana. 


elasticsearch {"@timestamp":"2026-04-05T17:25:09.928Z","log.level":"ERROR","message":"policy [jaeger-ilm-policy-archive] for index [jaeger-archive-jaeger-span-archive-000001] failed on step [{\"phase\":\"hot\",\"action\":\"rollover\",\"name\":\"check-r
ollover-ready\"}]. Moving to ERROR step", "ecs.version": "1.2.0","service.name":"ES_ECS","event.dataset":"elasticsearch.server","process.thread.name":"elasticsearch[elk-default-es-masters-0][trigger_engine_scheduler][T#1]","log.logger":"org.elasticsear
ch.xpack.ilm.IndexLifecycleRunner","elasticsearch.cluster.uuid":"4wGh1ggcSKSv8leROyWM2w","elasticsearch.node.id":"9G_7ARYYSV2f5oSZB4AD5A","elasticsearch.node.name":"elk-default-es-masters-0","elasticsearch.cluster.name":"elk-default","error.type":"java
.lang.IllegalArgumentException","error.message":"index.lifecycle.rollover_alias [jaeger-archive-jaeger-span-write] does not point to index [jaeger-archive-jaeger-span-archive-000001]","error.stack_trace":"java.lang.IllegalArgumentException: index.lifec
ycle.rollover_alias [jaeger-archive-jaeger-span-write] does not point to index [jaeger-archive-jaeger-span-archive-000001]\n\tat org.elasticsearch.xcore@9.3.2/org.elasticsearch.xpack.core.ilm.WaitForRolloverReadyStep.evaluateCondition(WaitForRolloverRe
adyStep.java:176)\n\tat org.elasticsearch.ilm@9.3.2/org.elasticsearch.xpack.ilm.IndexLifecycleRunner.runPeriodicStep(IndexLifecycleRunner.java:241)\n\tat org.elasticsearch.ilm@9.3.2/org.elasticsearch.xpack.ilm.IndexLifecycleService.triggerPolicies(Inde
xLifecycleService.java:535)\n\tat org.elasticsearch.ilm@9.3.2/org.elasticsearch.xpack.ilm.IndexLifecycleService.triggerPolicies(IndexLifecycleService.java:475)\n\tat org.elasticsearch.ilm@9.3.2/org.elasticsearch.xpack.ilm.IndexLifecycleService.triggere
d(IndexLifecycleService.java:456)\n\tat org.elasticsearch.server@9.3.2/org.elasticsearch.common.scheduler.SchedulerEngine.notifyListeners(SchedulerEngine.java:209)\n\tat org.elasticsearch.server@9.3.2/org.elasticsearch.common.scheduler.SchedulerEngine$
ActiveSchedule.run(SchedulerEngine.java:243)\n\tat java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:545)\n\tat java.base/java.util.concurrent.FutureTask.run(FutureTask.java:328)\n\tat java.base/java.util.concurrent.Scheduled
ThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:309)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1090)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPo
olExecutor.java:614)\n\tat java.base/java.lang.Thread.run(Thread.java:1474)\n"}
elasticsearch {"@timestamp":"2026-04-05T17:25:09.936Z","log.level":"ERROR","message":"policy [jaeger-ilm-policy-archive] for index [jaeger-archive-jaeger-span-write] failed on step [{\"phase\":\"hot\",\"action\":\"rollover\",\"name\":\"check-rollover-r
eady\"}]. Moving to ERROR step", "ecs.version": "1.2.0","service.name":"ES_ECS","event.dataset":"elasticsearch.server","process.thread.name":"elasticsearch[elk-default-es-masters-0][trigger_engine_scheduler][T#1]","log.logger":"org.elasticsearch.xpack.
ilm.IndexLifecycleRunner","elasticsearch.cluster.uuid":"4wGh1ggcSKSv8leROyWM2w","elasticsearch.node.id":"9G_7ARYYSV2f5oSZB4AD5A","elasticsearch.node.name":"elk-default-es-masters-0","elasticsearch.cluster.name":"elk-default","error.type":"java.lang.Ill
egalArgumentException","error.message":"index.lifecycle.rollover_alias [jaeger-archive-jaeger-span-write] does not point to index [jaeger-archive-jaeger-span-write]","error.stack_trace":"java.lang.IllegalArgumentException: index.lifecycle.rollover_alia
s [jaeger-archive-jaeger-span-write] does not point to index [jaeger-archive-jaeger-span-write]\n\tat org.elasticsearch.xcore@9.3.2/org.elasticsearch.xpack.core.ilm.WaitForRolloverReadyStep.evaluateCondition(WaitForRolloverReadyStep.java:176)\n\tat org
.elasticsearch.ilm@9.3.2/org.elasticsearch.xpack.ilm.IndexLifecycleRunner.runPeriodicStep(IndexLifecycleRunner.java:241)\n\tat org.elasticsearch.ilm@9.3.2/org.elasticsearch.xpack.ilm.IndexLifecycleService.triggerPolicies(IndexLifecycleService.java:535)
\n\tat org.elasticsearch.ilm@9.3.2/org.elasticsearch.xpack.ilm.IndexLifecycleService.triggerPolicies(IndexLifecycleService.java:475)\n\tat org.elasticsearch.ilm@9.3.2/org.elasticsearch.xpack.ilm.IndexLifecycleService.triggered(IndexLifecycleService.jav
a:456)\n\tat org.elasticsearch.server@9.3.2/org.elasticsearch.common.scheduler.SchedulerEngine.notifyListeners(SchedulerEngine.java:209)\n\tat org.elasticsearch.server@9.3.2/org.elasticsearch.common.scheduler.SchedulerEngine$ActiveSchedule.run(Schedule
rEngine.java:243)\n\tat java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:545)\n\tat java.base/java.util.concurrent.FutureTask.run(FutureTask.java:328)\n\tat java.base/java.util.concurrent.ScheduledThreadPoolExecutor$Schedule
dFutureTask.run(ScheduledThreadPoolExecutor.java:309)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1090)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:614)\n\tat
java.base/java.lang.Thread.run(Thread.java:1474)\n"}


But these are probably related to jaegertracing creating the `jaeger-archive-jaeger-span-write` as index instead of using it as alias I'd assume.
```

### Screenshot

Image

### Jaeger backend version

Version:  2.17.0

### Stogage backend

Elastic-Search 9.3.2

### Operating system

TalosOS 1.12.6

### Deployment model

OpenTelemetryCollector operator
Version: 0.148.0

### Deployment configs

```shell

```

### PR DESCRIPTION
**Pull request by vic-comm**
Fixes https://github.com/jaegertracing/jaeger/issues/8292

Description:
Resolves the ILM index collision bug where archive data overwrote primary storage. Parameterized the MappingBuilder to dynamically inject the -archive- segment into the rollover_alias of the Elasticsearch templates during initialization, aligning the JSON blueprint with the CLI output.

## Checklist
- [x] I have read https://github.com/jaegertracing/jaeger/blob/main/CONTRIBUTING_GUIDELINES.md
- [x] I have signed all commits
- [x] I have added unit tests for the new functionality
- [x] I have run lint and test steps successfully: `make lint test`

## AI Usage in this PR (choose one)
See [AI Usage Policy](https://github.com/jaegertracing/jaeger/blob/main/CONTRIBUTING_GUIDELINES.md#ai-usage-policy).
- [ ] **None**: No AI tools were used in creating this PR
- [ ] **Light**: AI provided minor assistance (formatting, simple suggestions)
- [x] **Moderate**: AI helped with code generation or debugging specific parts
- [ ] **Heavy**: AI generated most or all of the code changes

### CODE REVIEW

**Comment by vic-comm** [comment]
@yurishkuro, thanks for the earlier review, kindly add the changelog:bugfix label so the ci tests can pass

**Comment by vic-comm** [comment]
@yurishkuro, @jkowall kindly review and add the changelog:bugfix label. Thank you.

File: internal/storage/v1/elasticsearch/mappings/mapping.go [hunk 1]
```diff
@@ -34,6 +34,7 @@ type MappingBuilder struct {
 	EsVersion       uint
 	UseILM          bool
 	ILMPolicyName   string
+	Archive         bool
```
Comment (Manik2708) [inline internal/storage/v1/elasticsearch/mappings/mapping.go:37]: this means archive was not present at all in mappings that means all of the other parameters for example read alias should also be covered with archive. Is there any other parameter too? And as far I could see we have integration tests only for "init" action of rollover app and certainly harder to test this kind of edge case.

**Comment by Manik2708** [comment]
@vic-comm if possible can you please upload an image of manual verification of fix, if not then ping me, I will test whether this has fixed the bug or not

File: internal/storage/integration/es_index_rollover_test.go [hunk 2]
```diff
@@ -178,3 +178,60 @@ func cleanES(t *testing.T, client *elastic.Client, policyName string) {
...
+	indices, err := client.IndexNames()
+	require.NoError(t, err)
+	assert.Contains(t, indices, expectedIndex)
+
+	// Fetch the settings injected into the archive index
+	settings, err := client.IndexGetSettings(expectedIndex).FlatSettings(true).Do(context.Background())
+	require.NoError(t, err)
+
+	// Assert that the bug is fixed: The rollover alias MUST contain '-archive-'
+	require.Len(t, settings, 1)
```
Comment (Manik2708) [inline internal/storage/integration/es_index_rollover_test.go:230]: I maybe wrong but I don't think this assertion represents the actual issue. The problem was in mappings and we could see in the image that alias was created even then but traces were stored somewhere else. Probably due to mismatch from mappings, ES/OS themselves create the index and start storing into them. Conclusively this is not an alias creation problem but a direction problem (due to template mismatch it is directing towards a new index). Although I am confident that changes in mappings would have solved this issue but I feel a manual loop will be better to see if the bug is fixed or not.

Comment (vic-comm) [reply internal/storage/integration/es_index_rollover_test.go:230]: That is a fair point regarding the assertion. The core issue was ES auto-creating a raw index because the jaeger-span-archive-write alias was missing from the template injection.
I spun up a local ES 7 cluster and ran the init action manually using the compiled es-rollover tool with ARCHIVE=true.
As you can see in the screenshot below, the template mapping collision is resolved. The tool now successfully maps jaeger-span-archive-write to jaeger-span-archive-000001. Because this alias now exists natively, when the backend attempts to write an archived trace, ES will correctly route it to the rollover index rather than auto-generating a raw index.
Screenshot 2026-06-04 at 17 49 51

Comment (Manik2708) [reply internal/storage/integration/es_index_rollover_test.go:230]: have you sent some traces to jaeger? Because the core issue is not creation of alias but direction as specified in above comment!

Comment (vic-comm) [reply internal/storage/integration/es_index_rollover_test.go:230]: I see your point, we need to be sure ES respects the alias during a write instead of auto-creating a rogue index.
To verify the routing, I simulated a backend write directly to the jaeger-span-archive-write alias. As the screenshot shows, ES successfully routed the document to "_index":"jaeger-span-archive-000001". The alias routing works exactly as expected.
Screenshot 2026-06-04 at 17 49 51

Comment (vic-comm) [reply internal/storage/integration/es_index_rollover_test.go:230]: image

**Comment by Manik2708** [comment]
just look into read alias issue else lgtm

File: internal/storage/integration/es_index_rollover_test.go (hunk 2 above)
Comment (Manik2708) [reply internal/storage/integration/es_index_rollover_test.go:230]: then ok to write integration test but i think we can remove this comment

Comment (vic-comm) [reply internal/storage/integration/es_index_rollover_test.go:230]: Done

**Comment by vic-comm** [comment]
> just look into read alias issue else lgtm

No other parameters are missing. The JSON templates only require the rollover_alias. Read aliases are applied dynamically in action.go, which already handles the archive flag correctly

**Comment by Manik2708** [comment]
> > just look into read alias issue else lgtm
> 
> No other parameters are missing. The JSON templates only require the rollover_alias. Read aliases are applied dynamically in action.go, which already handles the archive flag correctly

How rollover alias was to be fixed and not read alias?

**Comment by vic-comm** [comment]
@Manik2708 JSON index templates only require the rollover_alias to direct ILM writes. Read aliases aren't defined in templates at all, they are created dynamically via API calls in action.go. Since the Archive flag is now plumbed through, that Go code automatically registers the correct jaeger-span-archive-read alias (which we verified in the cluster check earlier).

**Comment by Manik2708** [comment]
Another thing, we have to investigate for the users using archive storage and upgrading jaeger. Series of steps for them have to be listed, for example as per my observance users using ILM might have to run init again before upgrading and similar for users doing mannual rollover.

**Comment by Manik2708** [comment]
> @Manik2708 JSON index templates only require the rollover_alias to direct ILM writes. Read aliases aren't defined in templates at all, they are created dynamically via API calls in action.go. Since the Archive flag is now plumbed through, that Go code automatically registers the correct jaeger-span-archive-read alias (which we verified in the cluster check earlier).

https://github.com/jaegertracing/jaeger/blob/d1ba7a98e44f36debaf75d8b47ffaf0e1d03885f/internal/storage/v1/elasticsearch/mappings/jaeger-span-7.json#L5 Do we not require to change this?

**Comment by vic-comm** [comment]
> > @Manik2708 JSON index templates only require the rollover_alias to direct ILM writes. Read aliases aren't defined in templates at all, they are created dynamically via API calls in action.go. Since the Archive flag is now plumbed through, that Go code automatically registers the correct jaeger-span-archive-read alias (which we verified in the cluster check earlier).
> 
> https://github.com/jaegertracing/jaeger/blob/d1ba7a98e44f36debaf75d8b47ffaf0e1d03885f/internal/storage/v1/elasticsearch/mappings/jaeger-span-7.json#L5
> 
> Do we not require to change this?

I completely missed that. 
I just pushed an update adding the ReadAlias parameter to the templates alongside the RolloverAlias. Thank you for catching that! Let me know if everything looks solid now.
